### PR TITLE
[mlir][sparse] use shared value storage between wrapped iterator and the wrapper.

### PR DESCRIPTION
**Pull request by PeimingLiu**

### CODE REVIEW

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.cpp [hunk 1]
```diff
@@ -205,34 +205,48 @@ static Value offsetFromMinCrd(OpBuilder &b, Location l, Value minCrd,
...
 namespace {
 
-class TrivialIterator : public SparseIterator {
-  Value getLoopLo(OpBuilder &b, Location l) const {
-    // Dense loop are traversed by coordinate, delinearize the position to get
-    // the coordinate.
-    if (randomAccessible())
-      return SUBI(itPos, posLo);
-    return itPos;
+// The iterator that that traverse a concrete sparse tensor levels. High-level
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.cpp:208]: typo: that twice
typo: traverses (subject is singular)

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.cpp [hunk 2]
```diff
@@ -205,34 +205,48 @@ static Value offsetFromMinCrd(OpBuilder &b, Location l, Value minCrd,
...
   bool randomAccessible() const override { return isDenseLT(stl.getLT()); };
-  bool iteratableByFor() const override { return true; };
+  bool iteratableByFor() const override { return kind != IterKind::kDedup; };
   Value upperBound(OpBuilder &b, Location l) const override {
     return stl.size();
   };
 
+protected:
+  // Owner of the storage, all wrappers build on top of a concrete iterator
+  // shares the same storage such that the iterator values are always
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.cpp:236]: shares -> share (subject is plural)

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 3]
```diff
@@ -119,8 +132,8 @@ class SparseIterator {
   virtual Value upperBound(OpBuilder &b, Location l) const = 0;
 
   // Serializes and deserializes the current status to/from a set of values. The
-  // ValueRange should contain values that specifies the current postion and
-  // loop bound.
+  // ValueRange should contain values that are suffcient to recover the current
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:135]: typo: sufficient

also, sufficient for what?

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 4]
```diff
@@ -219,7 +243,16 @@ class SparseIterator {
   // For trivial iterators, it is the position; for dedup iterators, it consists
   // of the positon and the segment high, for non-empty subsection iterator, it
   // is the metadata that specifies the subsection.
-  MutableArrayRef<Value> itVals;
+  //
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:246]: remove leading 

//

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 5]
```diff
@@ -219,7 +243,16 @@ class SparseIterator {
...
+  // Note that the wrapped iterator shares the same storage to maintain itVals
+  // with it wrapper, which means the wrapped iterator might only owns a subset
+  // of all the values stored in itValStorage.
+  const unsigned itValsCnt;
+  SmallVectorImpl<Value> &itValsStorageRef;
+  // All other (loop invariant) values used by the iterator. Although these
+  // values are not updated between loop iteration, they still need to be passed
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:253]: between loop iterations (plural)

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 6]
```diff
@@ -132,10 +145,33 @@ class SparseIterator {
   };
   virtual void deserialize(ValueRange vs) { llvm_unreachable("unsupported"); };
 
+  //  virtual std::string getFuncNamePostfix() const = 0;
+  //  virtual SmallVector<Value> toFuncArgs() const = 0;
```
Comment (yinying-lisa-li) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:149]: Should the comments be removed?

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h (hunk 3 above)
Comment (PeimingLiu) [reply mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:135]: to recover the current iterator position as well as loop bound?

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 7]
```diff
@@ -132,10 +145,33 @@ class SparseIterator {
...
+  ValueRange forward(OpBuilder &b, Location l);
+
+  // Actual Implementation provided by derived class.
+  virtual void genInitImpl(OpBuilder &, Location, const SparseIterator *) = 0;
+  virtual ValueRange forwardImpl(OpBuilder &b, Location l) = 0;
+
+  // Returns a boolean value that equals `!it.end()`
+  virtual Value genNotEnd(OpBuilder &b, Location l) = 0;
+
+  // Dereference the iterator, loads the coordinate at the current position.
```
Comment (yinying-lisa-li) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:169]: dereferences for consistency.

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 8]
```diff
@@ -219,7 +242,16 @@ class SparseIterator {
   // For trivial iterators, it is the position; for dedup iterators, it consists
   // of the positon and the segment high, for non-empty subsection iterator, it
   // is the metadata that specifies the subsection.
-  MutableArrayRef<Value> itVals;
+  // Note that the wrapped iterator shares the same storage to maintain itVals
+  // with it wrapper, which means the wrapped iterator might only owns a subset
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:246]: typo: might only own (no s)

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h (hunk 3 above)
Comment (aartbik) [reply mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h]: Ah, I really meant, explain sufficient for what in the doc itself ;-)

File: mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h [hunk 9]
```diff
@@ -219,7 +242,16 @@ class SparseIterator {
...
+  // of all the values stored in itValStorage.
+  const unsigned itValsCnt;
+  SmallVectorImpl<Value> &itValsStorageRef;
+  // All other (loop invariant) values used by the iterator. Although these
+  // values are not updated between loop iterations, they still need to be
+  // passed as function parameters to reconstruct the iterator in a new function
+  // scope.
+  // SmallVectorImpl<Value> &metaValsStorageRef;
```
Comment (aartbik) [inline mlir/lib/Dialect/SparseTensor/Transforms/Utils/SparseTensorLevel.h:254]: remove dead code?

**Review by aartbik — APPROVED** [review]

**Review by yinying-lisa-li — APPROVED** [review]

**Pull request merged**